pkg/virt-handler/vm.go

```diff
@@ -986,6 +988,35 @@ func (d *VirtualMachineController) processVmCleanup(vmi *v1.VirtualMachineInstan
 	d.migrationProxy.StopTargetListener(string(vmi.UID))
 	d.migrationProxy.StopSourceListener(string(vmi.UID))
 
+	mountDir := containerdisk.GenerateVolumeMountDir(vmi)
+
+	files, err := ioutil.ReadDir(mountDir)
+	if err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("failed to list container disk mounts: %v", err)
+	}
+
+	if vmi.UID != "" {
+		// let containerdisks gracefully terminate if they still managed to run somehow
+		for _, file := range files {
+			path := filepath.Join(mountDir, file.Name())
+			if strings.HasSuffix(path, ".sock") {
+				continue
+			}
+			if mounted, err := isolation.NodeIsolationResult().IsMounted(path); err != nil {
+				return fmt.Errorf("failed to check mount point for containerDisk %v: %v", path, err)
+			} else if mounted {
+				out, err := exec.Command("/usr/bin/chroot", "--mount", "/proc/1/ns/mnt", "umount", path).CombinedOutput()
```

Comment:
what happens if the umount never occurs here? will the pod terminate still?

Reply:
The pod will terminate and disappears. But you will see that the kubelet periodically tries to unmount the containerDisk layer and it fails because another mount is still there. I can actually improve this by tracking in an extra file vms for which we did not unmount and periodically do the cleanup for these. This will allow to always clean up successfully as soon as virt-handler is started, no matter if VMIs disappeared or not in the meantime.

Reply:
> I can actually improve this by tracking in an extra file vms for which we did not unmount and periodically do the cleanup for these

doesn't watchdog file account for this though? we dont' delete the watchdog until the vmi is 100% cleaned up locally.

Reply:
yes you are right.